osu.Game/Screens/SelectV2/BeatmapCarouselFilterGrouping.cs

```diff
@@ -35,70 +36,100 @@ public BeatmapCarouselFilterGrouping(Func<FilterCriteria> getCriteria)
 
-            int starGroup = int.MinValue;
-
-            if (criteria.SplitOutDifficulties)
+            // Add criteria groups.
+            switch (criteria.Group)
             {
-                var diffItems = new List<CarouselItem>(items.Count());
-
-                GroupDefinition? group = null;
+                default:
+                    groupSetsTogether = true;
+                    newItems.AddRange(items);
+                    break;
 
-                foreach (var item in items)
-                {
-                    var b = (BeatmapInfo)item.Model;
+                case GroupMode.Difficulty:
+                    groupSetsTogether = false;
+                    int starGroup = int.MinValue;
 
-                    if (b.StarRating > starGroup)
+                    foreach (var item in items)
                     {
-                        starGroup = (int)Math.Floor(b.StarRating);
-                        group = new GroupDefinition($"{starGroup} - {++starGroup} *");
-                        diffItems.Add(new CarouselItem(group) { DrawHeight = GroupPanel.HEIGHT });
-                    }
+                        cancellationToken.ThrowIfCancellationRequested();
 
-                    if (!groupItems.TryGetValue(group!, out var related))
-                        groupItems[group!] = related = new HashSet<CarouselItem>();
-                    related.Add(item);
+                        var b = (BeatmapInfo)item.Model;
 
-                    diffItems.Add(item);
+                        if (b.StarRating > starGroup)
+                        {
+                            starGroup = (int)Math.Floor(b.StarRating);
+                            newItems.Add(new CarouselItem(new GroupDefinition($"{starGroup} - {++starGroup} *")) { DrawHeight = GroupPanel.HEIGHT });
+                        }
 
-                    item.IsVisible = false;
-                }
+                        newItems.Add(item);
+                    }
 
-                return diffItems;
+                    break;
             }
 
+            // Add set headers wherever required.
             CarouselItem? lastItem = null;
 
-            var newItems = new List<CarouselItem>(items.Count());
-
-            foreach (var item in items)
+            if (groupSetsTogether)
             {
-                cancellationToken.ThrowIfCancellationRequested();
-
-                if (item.Model is BeatmapInfo b)
+                for (int i = 0; i < newItems.Count; i++)
                 {
-                    // Add set header
-                    if (lastItem == null || (lastItem.Model is BeatmapInfo b2 && b2.BeatmapSet!.OnlineID != b.BeatmapSet!.OnlineID))
+                    cancellationToken.ThrowIfCancellationRequested();
+
+                    var item = newItems[i];
+
+                    if (item.Model is BeatmapInfo beatmap)
                     {
-                        newItems.Add(new CarouselItem(b.BeatmapSet!)
+                        if (groupSetsTogether)
```

Comment:
isn't this already checked in the outermost conditional? (line 80)

Reply:
Removed this.